From 934bacb7786eb353cf74e1d58d9f7226e6f6297c Mon Sep 17 00:00:00 2001
From: Timon Sassor <timon.sassor@governikus.de>
Date: Wed, 30 Jul 2025 12:44:36 +0200
Subject: Android: Use the TalkBack focus signal to trigger focusAction

TalkBack provides A11y specific focus signals that are emitted when
the focus frame is moved. It would be useful to pass this focus
change to the focused property of an item.

[ChangeLog][Accessibility][Android] Use TalkBack
ACTION_ACCESSIBILITY_FOCUS to trigger focusAction.

Fixes: QTBUG-137769
Pick-to: 6.10 6.9 6.8
Change-Id: I63219cfb9b8776ddaa6c85ccf1844adb1984d2b7
Reviewed-by: Assam Boudjelthia <assam.boudjelthia@qt.io>
(cherry picked from commit e7a211d702d6aa21a30eff56ceeffc153385443e)
---
 .../qt/android/QtAccessibilityDelegate.java   |  3 ++
 .../qt/android/QtNativeAccessibility.java     |  1 +
 .../android/androidjniaccessibility.cpp       | 28 +++++++++++++++++++
 3 files changed, 32 insertions(+)

diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
index 6c0bd5d75eccfc402035e6f5139b3614185af549..bc19882840d74ed2b8a6cb40323ed73508b1a6ce 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
@@ -569,6 +569,9 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
             if (success)
                 sendEventForVirtualViewId(virtualViewId, AccessibilityEvent.TYPE_VIEW_CLICKED);
             break;
+        case AccessibilityNodeInfo.ACTION_ACCESSIBILITY_FOCUS:
+            success = QtNativeAccessibility.focusAction(virtualViewId);
+            break;
         case AccessibilityNodeInfo.ACTION_SCROLL_FORWARD:
             success = QtNativeAccessibility.scrollForward(virtualViewId);
             if (success)
diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
index dd2cead8cd0c7d3c1fcd86508620749303cf2747..e63de6f79756198f1f3de1e4fc894c6a4748d2c4 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
@@ -15,6 +15,7 @@ class QtNativeAccessibility
     static native Rect screenRect(int objectId);
     static native int hitTest(float x, float y);
     static native boolean clickAction(int objectId);
+    static native boolean focusAction(int objectId);
     static native boolean scrollForward(int objectId);
     static native boolean scrollBackward(int objectId);
 
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index 88db149ed012d9ed5c037765ad8629ed25e71157..a88bfc00eaf57f995465023d55560410c8f61495 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -316,6 +316,22 @@ namespace QtAndroidAccessibility
         return true;
     }
 
+    static bool focusAction_helper(int objectId)
+    {
+        QAccessibleInterface *iface = interfaceFromId(objectId);
+        if (!iface || !iface->isValid() || !iface->actionInterface())
+            return false;
+
+        const auto& actionNames = iface->actionInterface()->actionNames();
+
+        if (actionNames.contains(QAccessibleActionInterface::setFocusAction())) {
+            invokeActionOnInterfaceInMainThread(iface->actionInterface(),
+                                                QAccessibleActionInterface::setFocusAction());
+            return true;
+        }
+        return false;
+    }
+
     static jboolean clickAction(JNIEnv */*env*/, jobject /*thiz*/, jint objectId)
     {
         bool result = false;
@@ -327,6 +343,17 @@ namespace QtAndroidAccessibility
         return result;
     }
 
+    static jboolean focusAction(JNIEnv */*env*/, jobject /*thiz*/, jint objectId)
+    {
+        bool result = false;
+        if (m_accessibilityContext) {
+            runInObjectContext(m_accessibilityContext, [objectId]() {
+                return focusAction_helper(objectId);
+            }, &result);
+        }
+        return result;
+    }
+
     static bool scroll_helper(int objectId, const QString &actionName)
     {
         QAccessibleInterface *iface = interfaceFromId(objectId);
@@ -719,6 +746,7 @@ namespace QtAndroidAccessibility
         {"hitTest", "(FF)I", (void*)hitTest},
         {"populateNode", "(ILandroid/view/accessibility/AccessibilityNodeInfo;)Z", (void*)populateNode},
         {"clickAction", "(I)Z", (void*)clickAction},
+        {"focusAction", "(I)Z", (void*)focusAction},
         {"scrollForward", "(I)Z", (void*)scrollForward},
         {"scrollBackward", "(I)Z", (void*)scrollBackward},
     };
